### ISSUE TITLE
Eventually should also catch AssertionError so that you can test assertions within

### ISSUE THREAD

**Issue by JLLeitschuh** [kotlintest/kotlintest#145]
If an assertion error (as all of these tests matchers throw) is thrown within the `eventually` then it immediately exits without running again.

https://github.com/kotlintest/kotlintest/blob/master/src/main/kotlin/io/kotlintest/Eventually.kt#L14

**Comment by sksamuel** [comment]
If it catches it and returns it then it wouldn't be able to be caught by the framework right?

**Comment by aleixmorgadas** [comment]
I've made some tests to view the actual behavior.
```kotlin
"throws AssertionError" {
    val start = System.nanoTime()
    eventually(5000.microseconds) {
        (1 .. 10000).forEach({5+5+5+5+5+5})
    }
    val end = System.nanoTime()
    println("time elapsed ${(end- start)*1000} micro sec")
}
```
**Result**: Test OK . `time elapsed 3334420000 micro sec`

AFAIK, the implementation does not achieve the expected behavior. Taking into account the [documentation](https://github.com/kotlintest/kotlintest/blob/release/2.0/doc/reference.md#eventually-)
> Eventually:
> When testing future based code, [...]

I suggest enforce  `eventually` function signature to be Future-based. Like:
```kotlin
fun <T>eventually(duration: Duration, f: () -> Future<T>): T
```

What do you think?

**Comment by sksamuel** [comment]
Eventually is blocking, so it doesn't necessarily have to return a future. In fact if it does, what's the real point of it.

**Comment by aleixmorgadas** [comment]
Understood. Then, is test result the expected?

I've found another code that I think that it behaves wrongly.
```kotlin
"find the 1000 number" {
    val iterator = (1..1000).iterator()
    eventually(2.seconds) {
        iterator.next() shouldBe 1000
    }
}
```
**Result**: Test fails . `java.lang.AssertionError: 1 did not equal 1000`

Because it eventually will be 1000, the test should pass, right?

**Comment by aleixmorgadas** [comment]
Understood. Then, is test result the expected?

I've found another code that I think that it behaves wrongly.
```kotlin
"find the 1000 number" {
    val iterator = (1..1000).iterator()
    eventually(2.seconds) {
        iterator.next() shouldBe 1000
    }
}
```
**Result**: Test fails . `java.lang.AssertionError: 1 did not equal 1000`

Because it eventually will be 1000, the test should pass, right?

**Comment by sksamuel** [comment]
Yes it should pass.

**Comment by sanity** [comment]
FYI - I created my own version of this.  It doesn't have a "busy" while loop, and it remembers the last exception to be thrown by f() and includes it in the `AssertionError` thrown when the maximumTime is exceeded.

```kotlin
val Duration.millis get() = this.timeUnit.toMillis(amount)
fun <T> ians_eventually(maximumTime: io.kotlintest.Duration, pollEvery : Duration = 100.milliseconds, f: () -> T): T {
    val end = System.nanoTime() + maximumTime.nanoseconds
    var times = 0
    var lastException : Exception? = null
    while (System.nanoTime() < end) {
        try {
            return f()
        } catch (e: Exception) {
            lastException = e
        }
        Thread.sleep(pollEvery.millis)
        times++
    }
    throw AssertionError("Test failed after ${maximumTime.amount} ${maximumTime.timeUnit}; attempted $times times", lastException!!)
}
```

**Comment by sanity** [comment]
Also, perhaps off-topic, but why is kotlintest using it's own custom `Duration` rather than a `java.time.Duration`?

**Comment by helmbold** [comment]
Because java.time.Duration is not available for Android developers, which are still not on Java 8.

**Comment by sanity** [comment]
Ah, of course - makes sense.  





On Mon, May 8, 2017 7:46 AM, Christian Helmbold 'notifications@github.com' via
33Mail sender@mailer1.33mail.com  wrote:
Because java.time.Duration is not available for Android developers, which are
still not on Java 8.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.


This email was sent to the alias 'github@ian.33mail.com' by
'reply@reply.github.com',
and 33Mail forwarded it to you. To block all further emails to this alias click 
here  
(Prefer no ads? Upgrade to Premium.)  

Ian Clarkehttp://blog.locut.us/
